Contributions/Linux_Memory_Management_Essentials.md

```diff
@@ -0,0 +1,249 @@
+         only when written to, then a separate physical page is reserved for each writing process (threads sahre the same)
+      2. same happens for data pages that were treated as zero-page, but then are written to.
+   5. folios (see also kernel section): data structures that try to better abstract compound pages and *might* also be used
+      to represent optimised contiguous pages on ARM64 (instead of mapping 16 entries, it is possible
+      to map a 16-pages chunk of physically contiguous memory, that is also aligned to a 16 pages
+      boundary, reducing TLB use).
+      1. A folio acts as intermediary between vma and lower level memory management
+      2. it might pre-allocate/map more pages than explicitly requested
+7. read-ahead: when asked to fetch data from disk, the kernel might attempt to optimise the operation,
+   reading more pages than requested, under the assumption that more requests might be coming soon
+ 
+
+#### **Safety-Oriented consideration**
+The following considerations are of a more deductive nature.
+
+1. a process that is supposed to support safety requirements should  not have pages swapped out / dropped / missing,
+   because this would introduce:
+   1. uncertainty in the timing required to recover the content, if not immediately available
+   2. additional risk, involving the userspace paging mechanisms in the fulfilling of the safety requirements
+   3. additional dependency on runtime linking, in case the process requires it, and code pages have been
+      discarded - reloading them from disk will not be sufficient
+2. The optimisations made by the kernel in providing physical backing to process memory make it very
+   questionable if it can be assessed when a (part of) a process memory content is actually present in the
+   system physical memory.
+3. by default, it is to be expected that a process will be exposed to various types of interference from the kernel:
+   1. some of a more bening nature, like dropping of pages, or not allocation of not-yet-used one
+   2. some limited in extent, but hard or even practicaly impossible to detect, like a rogue write to process physical memory
+   3. some of systemic nature, like some form of use-after free, where a process page is accidentally in use also by another component
+   4. some of indirect nature, like for example when the page table of the process address space is somehow corrupted
+4. again, because of the extremely complex nature of the system, positive testing is not sufficient, but it needs to
+   be paired also with negative testing, proving that it is possible to cope with interference and detect it, somehow.
+5. the same considerations made about integrity vs. avaialbility for the kernel are valid here too: detecting
+   interference doesn't help with keeping it under a certain threshold, and due to the complexity of the system,
+   it is not possible to estimate the risk reliably.
```

Comment:
Estimate which risk? Availability risk ?

Reply:
risk of lack of availability, but yes

Reply:
closing as outdated